src/shared/mount-util.c

```diff
@@ -742,3 +745,228 @@ int mount_option_mangle(
 
         return 0;
 }
+
+int bind_mount_in_namespace(
+                pid_t target,
+                const char *propagate_path,
+                const char *incoming_path,
+                const char *src,
+                const char *dest,
+                int read_only,
+                int make_file_or_directory) {
+
+        _cleanup_close_pair_ int errno_pipe_fd[2] = { -1, -1 };
+        char mount_slave[] = "/tmp/propagate.XXXXXX", *mount_tmp, *mount_outside, *p;
+        bool mount_slave_created = false, mount_slave_mounted = false,
+                mount_tmp_created = false, mount_tmp_mounted = false,
+                mount_outside_created = false, mount_outside_mounted = false;
+        _cleanup_free_ char *chased_src = NULL;
+        struct stat st;
+        pid_t child;
+        int r;
+
+        assert(target > 0);
+        assert(propagate_path);
+        assert(incoming_path);
+        assert(src);
+        assert(dest);
+
+        /* One day, when bind mounting /proc/self/fd/n works across
+         * namespace boundaries we should rework this logic to make
+         * use of it... */
+
+        p = strjoina(propagate_path, "/");
+        if (laccess(p, F_OK) < 0)
+                return log_debug_errno(SYNTHETIC_ERRNO(r == -ENOENT ? -EOPNOTSUPP : r), "Target does not allow propagation of mount points");
+
+        r = chase_symlinks(src, NULL, CHASE_TRAIL_SLASH, &chased_src, NULL);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to resolve source path of %s: %m", src);
+
+        if (lstat(chased_src, &st) < 0)
+                return log_debug_errno(errno, "Failed to stat() resolved source path %s: %m", chased_src);
+        if (S_ISLNK(st.st_mode)) /* This shouldn't really happen, given that we just chased the symlinks above, but let's better be safe… */
+                return log_debug_errno(SYNTHETIC_ERRNO(EOPNOTSUPP), "Source directory %s can't be a symbolic link", chased_src);
```

Comment:
(btw, if you want to make this code nicer, you could use the last argument to chase_symlinks(), and acquire an O_PATH fd of the target, then use fstat() on that, and then use /proc/self/fd/… on that very fd in the mount calls below (which should then use mount_follow_verbose() though, given you don't want the /proc/self/fd dance twice). That would be super sexy, because suddenly race free, as we'd always operate on pinned inodes rather than paths

Reply:
(but might be something for a later PR, dunno, up to you. This is just code shifted around after all, not new code

Reply:
sounds like a great idea, added a new commit - PTAL!